### PR TITLE
Fugw loss on val features

### PR DESCRIPTION
**Pull request by pbarbarant**
Small changes to compute the fugw loss on validation features :

- Adds the ability to compute the loss at each BCD iteration on a validation set comprised of feature maps/anatomical kernels (or both);
- New unit tests;
- Warnings and error messages when lacking or incorrect data is passed.

### CODE REVIEW

File: src/fugw/mappings/dense.py [hunk 1]
```diff
@@ -125,6 +143,23 @@ def fit(
         Ds = _make_tensor(source_geometry, device=device)
         Dt = _make_tensor(target_geometry, device=device)
 
+        # Do the same for validation data if it was provided
+        if source_features_val is not None and target_features_val is not None:
+            Fs_val = _make_tensor(source_features_val.T, device=device)
+            Ft_val = _make_tensor(target_features_val.T, device=device)
+            F_val = torch.cdist(Fs_val, Ft_val, p=2) ** 2
+
```
Comment (alexisthual) [inline src/fugw/mappings/dense.py:151]: I like it better when `if, elif, else` statements are grouped together, but that's just a personal preference ;)

**Review by alexisthual — COMMENTED** [review]
This is going in the right direction!

However, I think it'd make more sense to have `F_val` and `Ds_val`, `Dt_val` be non-null if and only if at least one of the two couples `[source_features_val, target_features_val]`, `[source_geometry_val, target_geometry_val]` has 2 non-null values.
If I remember correctly, computing a fugw loss takes a bit of time, and validation data will definitely eat up GPU space, so it'd be a shame to compute the validation loss if it does not bring new information.
This would imply changes in `{mappings,solvers}/{dense,sparse}.py`

Moreover, I am not sure I see what would be a good use case for setting `source_geometry_val` and `target_geometry_val`, although they might end up taking a lot of memory space . Can you think of one? :thinking: 

Besides, we are still missing tests :slightly_smiling_face: 
In particular, I would test that:
- things run smoothly with and without validation data (no validation data, validation features only, validation geometries only, both)
- we raise useful errors when validation data exist for source and not target (and vice-versa)
- returned validation losses match training validation losses when validation data = training data

File: src/fugw/mappings/dense.py [hunk 2]
```diff
@@ -149,17 +150,53 @@ def fit(
...
+            )
+
         else:
             F_val = None
 
+            # Raise warning if validation feature maps are not provided
+            warnings.warn(
+                "Validation data for feature maps is not provided."
+                " Using training data instead."
+            )
```
Comment (alexisthual) [inline src/fugw/mappings/dense.py:172]: I don't think we should raise a warning (it will be printed in a majority of cases, although there is no real problem).
Instead, we could log this message if `verbose=True`. You can import fugw's rich console with `from fugw.utils import console` and use it this way:
```suggestion
            console.log(
                "Validation data for feature maps is not provided."
                " Using training data instead."
            )
```

File: src/fugw/mappings/dense.py [hunk 3]
```diff
@@ -149,17 +150,53 @@ def fit(
...
+
         else:
             Ds_val = None
             Dt_val = None
 
+            # Raise warning if validation anatomical kernels are not provided
+            warnings.warn(
+                "Validation data for anatomical kernels is not provided."
+                " Using training data instead."
+            )
```
Comment (alexisthual) [inline src/fugw/mappings/dense.py:198]:
```suggestion
            console.log(
                "Validation data for anatomical kernels is not provided."
                " Using training data instead."
            )
```

File: src/fugw/mappings/sparse.py [hunk 4]
```diff
@@ -168,9 +168,27 @@ def fit(
...
+            )
+
         else:
             F1_val, F2_val = None, None
 
+            # Raise warning if validation feature maps are not provided
+            warnings.warn(
+                "Validation data for feature maps is not provided."
+                " Using training data instead."
+            )
```
Comment (alexisthual) [inline src/fugw/mappings/sparse.py:190]:
```suggestion
            console.log(
                "Validation data for feature maps is not provided."
                " Using training data instead."
            )
```

File: src/fugw/mappings/sparse.py [hunk 5]
```diff
@@ -186,10 +204,34 @@ def fit(
...
+
         else:
             Ds1_val, Ds2_val = Ds1, Ds2
             Dt1_val, Dt2_val = Dt1, Dt2
 
+            # Raise warning if validation anatomical kernels are not provided
+            warnings.warn(
+                "Validation data for anatomical kernels is not provided."
+                " Using training data instead."
+            )
```
Comment (alexisthual) [inline src/fugw/mappings/sparse.py:233]:
```suggestion
            console.log(
                "Validation data for anatomical kernels is not provided."
                " Using training data instead."
            )
```

**Review by alexisthual — COMMENTED** [review]
Almost there!
I have a minor comment on how to log warnings.

File: examples/03_validation_set/plot_1_simple_crossval.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+"""
+==================================================================
+Using a validation set to monitor the convergence of the fugw loss
+==================================================================
+
+In this example, we use 3 fMRI feature maps for training and 2 independant
+fMRI feature maps for testing to examine the evolutions of a training and
+a validation loss on 2 low-resolution brain volumes.
+"""
+# sphinx_gallery_thumbnail_number = 9
```
Comment (alexisthual) [inline examples/03_validation_set/plot_1_simple_crossval.py:11]: This determines which image generated in this example gets to be the thumbnail in the example gallery.
I think the second image (showing the 2 losses) is the most interesting one :)
```suggestion
# sphinx_gallery_thumbnail_number = 2
```

File: examples/03_validation_set/plot_1_simple_crossval.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+# Plot the evolution of losses on train and test datasets.
+fig, ax1 = plt.subplots()
+
+color = "tab:blue"
+ax1.set_xlabel("BCD Step")
+ax1.set_ylabel("FUGW loss train", color=color)
+ax1.plot(mapping.loss_steps, mapping.loss["total"], color=color)
+ax1.tick_params(axis="y", labelcolor=color)
+
+ax2 = ax1.twinx()  # instantiate a second axes that shares the same x-axis
```
Comment (alexisthual) [inline examples/03_validation_set/plot_1_simple_crossval.py:127]: I'm a bit picky here, but the twin axes make it harder to compare the training and test losses (because they no longer have the same y-axis range).
It will be useful for plotting the cross-validation correlation though.

File: examples/03_validation_set/plot_1_simple_crossval.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+mapping = FUGW(alpha=0.5, rho=1, eps=1e-4)
+_ = mapping.fit(
+    source_features=source_features_normalized[:n_training_contrasts],
+    target_features=target_features_normalized[:n_training_contrasts],
+    source_geometry=source_geometry_normalized,
+    target_geometry=target_geometry_normalized,
+    source_features_val=source_features_normalized[n_training_contrasts:],
+    target_features_val=target_features_normalized[n_training_contrasts:],
+    source_geometry_val=source_geometry_normalized,
+    target_geometry_val=target_geometry_normalized,
```
Comment (alexisthual) [inline examples/03_validation_set/plot_1_simple_crossval.py:109]: Maybe you should say a word about the fact that `{source,target}_geometry_val` are given the same values as `{source,target}_geometry`.
`{source,target}_geometry_val` could be omitted here, and maybe you could say that it's hard to find a brain-related example where these should be different at validation time than at training time.

File: tests/mappings/test_dense_mapping.py [hunk 9]
```diff
@@ -88,3 +88,100 @@ def test_dense_mapping(device, return_numpy, solver):
...
+            source_geometry=source_geometry,
+            target_geometry=target_geometry,
+            solver="sinkhorn",
+            solver_params={
+                "nits_bcd": 3,
+                "ibpp_eps_base": 1e8,
+            },
+            device="cpu",
+        )
+        assert len(fugw.loss_val) == len(fugw.loss)
```
Comment (alexisthual) [inline tests/mappings/test_dense_mapping.py:134]: In this case, the returned validation loss should be equal to that of training.
I think the following line should test deep equality:
```suggestion
        assert fugw.loss_val == fugw.loss
```

File: tests/mappings/test_sparse_mapping.py [hunk 10]
```diff
@@ -157,3 +157,93 @@ def test_fugw_sparse_with_init(device, sparse_layout, return_numpy):
...
+    if validation == "None":
+        fugw.fit(
+            source_features=source_features_train,
+            target_features=target_features_train,
+            source_geometry_embedding=source_embeddings,
+            target_geometry_embedding=target_embeddings,
+            init_plan=init_plan,
+            device=device,
+        )
+        assert len(fugw.loss_val) == len(fugw.loss)
```
Comment (alexisthual) [inline tests/mappings/test_sparse_mapping.py:208]:
```suggestion
        assert fugw.loss_val == fugw.loss
```

**Review by alexisthual — COMMENTED** [review]
One last round of comments and I think we are good to go :)

**Review by alexisthual — APPROVED** [review]
LGTM, thanks for all these changes :)
